### PR TITLE
Remove recursion in ForAttributeWithMetadataName

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/70790.

@dotnet/roslyn-compiler ptal

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests_Attributes_FullyQualifiedName.cs [hunk 1]
```diff
@@ -1304,6 +1304,53 @@ static void replace(ref Compilation compilation, CSharpParseOptions parseOptions
         }
     }
 
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/pull/70791")]
+    public void FindCorrectAttributeOnTopLevelClass_WhenSearchingForClassDeclaration1_DeepRecursion()
+    {
+        var deeplyRecursive = string.Join("+", Enumerable.Repeat(""" "a" """, 20_000));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests_Attributes_FullyQualifiedName.cs:1310]: test fails at around 11,000 frames on my machine.

**Comment by CyrusNajmabadi** [comment]
@chsienki @333fred ptal

**Comment by ToddGrun** [comment]
I closed out [a PR](https://github.com/dotnet/roslyn/pull/70792) that solved this slightly differently. I used GreenNode.EnumerateNodes which led to a simple solution, but Cyrus pointed out an issue with using it in that it returns an IEnumerable which leads to boxing and vcalls on the enumerator. As that method is internal, I think we should consider having it return a concrete type and utilize it to simplify this change. Of course, the last time I made a change to fix up boxing enumerables in the compiler, things didn't go so well...

File: src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs [hunk 2]
```diff
@@ -69,23 +71,37 @@ internal abstract class AbstractSyntaxHelper : ISyntaxHelper
...
+            try
             {
-                var child = node.GetSlot(i);
+                stack.Push(root);
 
-                if (child is null || child.IsToken)
-                    continue;
+                while (stack.Count > 0)
+                {
+                    var node = stack.Pop();
```
Comment (333fred) [inline src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs:83]:
```suggestion
                while (stack.TryPop(out var node))
                {
```

File: src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs [hunk 3]
```diff
@@ -43,6 +43,8 @@ internal interface ISyntaxHelper
 
     internal abstract class AbstractSyntaxHelper : ISyntaxHelper
     {
+        private static readonly ObjectPool<Stack<GreenNode>> s_nodeStackPool = new ObjectPool<Stack<GreenNode>>(static () => new Stack<GreenNode>());
```
Comment (333fred) [inline src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs:46]: Consider just using an `ArrayBuilder`; they're already pooled, and we use them as stacks all over the place. We even have a nice `TryPop` extension for them :slightly_smiling_face:.

**Review by 333fred — COMMENTED** [review]
Done review pass

File: src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs:46]: Add and TryPop, name a more iconic duo :-D

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs:46]: Will make those changes. Thanks!

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests_Attributes_FullyQualifiedName.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests_Attributes_FullyQualifiedName.cs:1310]: Should this go in EndToEndTests.cs, with the rest of the deeply-recursive tests?

File: src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs]: done.

File: src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SourceGeneration/ISyntaxHelper.cs]: done.

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests_Attributes_FullyQualifiedName.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests_Attributes_FullyQualifiedName.cs]: Moved test.

**Review by jjonescz — APPROVED** [review]

**Pull request merged**